Make xfrd keep valid zone transfers around until reload succeeds

k0ekk0ek opened the pull request:
This PR prepares NSD for zone verification (previously a separate product called CreDNS) to be merged back. Summary of changes:

* Make configuration options context aware. This allows the same option name to be used for different configuration sections.
* Simplify and update initial socket initialization. This was necessary to have optional IPv6 sockets for both normal operation and verification.
* Set socket options, e.g. reuseaddr and non-blocking operation in separate functions. This makes the code a lot more readable and allows for reuse of the logic when creating the sockets for verification.
* Always use the recvmmsg and sendmmsg code paths. This allows for the elimination of parallel code paths and allows for reuse of the logic for adding handlers.
* Update xfrd to keep zone transfers around until reload succeeds so that updates do not have to fetched again if verification for a single zone fails.

As indicated in the title, it's still a work in progress, but initial review comments have been processed and integrated back into the commits to which they applied.

The last step will be the addition of the code that actually calls the verifier (apart from possibly having to process review comments of course)

Commit:
Make configuration parser context aware

michaelforney committed:
Avoid pointer arithmetic on `void *`

The pointer operand to the binary `+` operator must be to a complete
object type.

michaelforney committed:
Avoid unnecessary VLA

hdais committed:
minimise-any: prefer polular and small rrsets

Commit:
Ignore Eclipse IDE Project Settings

Commit:
Always use recvmmsg and sendmmsg

Eliminate multiple UDP handler implementations by always using the recvmmsg and
sendmmsg paths and provide compatible implementations if the operating system
does not or --enable-recvmmsg was not specified.

Commit:
Set socket options in separate functions

Improve readability and code-reuse by setting socket options in seperate
functions and eliminating the socket copying logic for non-udp sockets if
reuseport is specified.

Commit:
Simplify and correct initial socket setup

Commit:
Move tests from tpkg archives to tdir directories

Store tests in tdir directories as used by unbound to eliminate binary blobs
from the repository and allow Git to keep track of changes.

Commit:
Update xfrd to keep zone transfers until reload succeeds

Keep track of individual zone transfers so that correct updates can be
reapplied if a reload is unsuccessful.

wcawijngaards committed:
- Fix #38: log address and failure reason with tls handshake errors,
  squelches (the same as unbound) some unless high verbosity is used.

wcawijngaards committed:
- Fixup clang analysis warning in xfrd_parse_received_xfr_packet
  master dereference.

wcawijngaards committed:
Another fix for clang analysis.

wcawijngaards commented on tpkg/ixfr_faildel.tdir/ixfr_faildel.test:
```diff
@@ -7,8 +7,6 @@
 PRE="../../.."
 DIG=dig
 
-sleep 6
```
Maybe best to keep this for slow executing test hosts.

k0ekk0ek replied:
Fixed.

wcawijngaards commented on xfrd.c:
```diff
@@ -1191,86 +1206,63 @@ xfrd_set_timer(xfrd_zone_type* zone, time_t t)
-			zone->soa_notified_acquired = 0;
-		}
-		if(zone->soa_notified_acquired && zone->state == xfrd_zone_ok)
+		assert(serial != 0);
```
This could be 0 right? In an actual zone transfer, 0 serial is possible in a SOA record.

wcawijngaards commented on xfrd.c:
```diff
@@ -1191,86 +1206,63 @@ xfrd_set_timer(xfrd_zone_type* zone, time_t t)
+		assert(state == xfrd_xfr_corrupt ||
+		       state == xfrd_xfr_inconsistent);
+		while(xfr && 0 != compare_serial(ntohl(xfr->soa.serial),
+		                                 ntohl(serial)))
```
The call to compare_serial() is not needed, for inequality test.

wcawijngaards commented on xfrd.c:
```diff
@@ -1965,16 +1986,15 @@ xfrd_parse_received_xfr_packet(xfrd_zone_type* zone, buffer_type* packet,
-						       "update indicating "
-						       "current serial",
+		} else if(compare_serial(ntohl(soa->serial),
+		                         ntohl(last_soa->serial)) == 0)
```
Also this call to compare_serial() is not needed.  It could be: ntohl(soa->serial) == ntohl(..)

wcawijngaards commented on xfrd.c:
```diff
@@ -1965,16 +1986,15 @@ xfrd_parse_received_xfr_packet(xfrd_zone_type* zone, buffer_type* packet,
-			if(zone->soa_nsd.serial == soa->serial)
-				zone->soa_nsd_acquired = xfrd_time();
+			*last_soa_acquired = xfrd_time();
+			/* FIXME: zone state should not be updated here(?) */
```
Yes the zone state should be updated because we got a new transfer indicating the serial.  Hence we got a new lease time for the serial and thus the timestamp that denotes the leasetime is updated and also the zone state can be changed from (whatever, expired, retrying) to ok.  This means the FIXME comment can be removed, I think?

wcawijngaards replied:
And also the removed line that updates the soa_nsd_acquired if it also needs updating for the timestamp is needed.  Because the zone may be loaded and that zone has a new lease, is more recent, right?

k0ekk0ek replied:
Better safe than sorry I guess. Removed the comment. As for "the removed line": `last_soa_acquired` points to either the acquired for the transfer or the one that's loaded. Depending on whether there's a transfer or not.

wcawijngaards commented on xfrd.c:
```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+				break;
 			}
 		}
+		/* refresh if no full transfer was awaiting scheduling */
```
At this point the comment says you look at if a full transfer was waiting, but the code did not check for that.  It removed ixfrs, but did not check if there is an AXFR or if the loop ended nonNULL.

k0ekk0ek replied:
Correct. I had that in there at some point (comment probably remained), but (from memory) `msg_is_ixfr` might be certain, it didn't seem to be that if that's `false` it's necessarily an axfr. Just doing a refresh in all scenarios seemed like a better decision. Of course, I might have overlooked something or just doing the wrong check all together.

wcawijngaards replied:
It may be required to do a full AXFR after DNSSEC corrupt contents.  Otherwise the master may reply with a smaller incremental update that moves to the failed position every time, it would have to reply with a concatenation or compaction of all updates so that when the zone DNSSEC failure is solved we actually get it.  Probably IXFR is good enough but it depends on the master service of IXFR, right now we rely on the master sending an IXFR update to the latest version, and I think that is actually OK.

wcawijngaards commented on xfrd.c:
```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+			continue; /* continue if nothing has been committed */
+		}
+
+		if((zone->soa_notified_acquired != 0 &&
+		    compare_serial(ntohl(zone->soa_notified.serial),
+		                   ntohl(zone->soa_nsd.serial)) <= 0) ||
+		   (zone->soa_notified_acquired < zone->soa_nsd_acquired))
```
This is a && b || c without brackets, what do you mean?  Also if a notify for a later serial arrives before you process the earlier serial, I would think that soa_notify_acquired (time_t) < nsd_acquired (time_t) but we do not want to just ignore that notification, instead act on it to fetch that serial number.

wcawijngaards replied:
Wait now I see (a&&b) brackets, so the comment is about the time comparison part and ignoring earlier notifications.

wcawijngaards commented on xfrd.c:
```diff
@@ -2383,78 +2525,255 @@ find_same_master_notify(xfrd_zone_type* zone, int acl_num_nfy)
+		   individually. still, the logic is off if a completed
+		   transfer is around after the reload */
+		//assert(zone->soa_disk_acquired &&
+		//       zone->soa_disk_acquired <= xfrd->reload_cmd_last_sent);
```
Yes, and I would like an assertion for that.  Also, the `//` operand is imported from C++ and not in early C standards, so I think we should not use that, that improves portability.

k0ekk0ek replied:
Reinstated.

Commit:
Enable close-on-exec flag for sockets

Commit:
Add popen3

Commit:
Add zone resource record iterator

Commit:
Add option to disable additional out-of-zone data

Add option to not return additional information from disparate zones.

Commit:
Fix issues with confine-to-zone option boolean change

Fixes a logical typo and the interpretation of the dname_compare result

wcawijngaards committed:
- For #21 add
  contrib/patch_for_s6_startup_and_other_service_supervisors.diff
  that adds support for readiness notification with READY_FD from
  Cameron Nemo.

wcawijngaards committed:
Merge branch 'master' of github.com:NLnetLabs/nsd

Commit:
Make checkconf test accept confine-to-zone

wcawijngaards committed:
Fixup compile issues.

wcawijngaards committed:
Remove double free.

wcawijngaards committed:
Fixup region cleanups for socket allocation.

wcawijngaards committed:
And use tdir for generating reports, so it works with mini_tdir.

Commit:
Add changelog entries and release notes

Commit:
Merge pull request #41 from NLnetLabs/fixup-compile

Fixup compile issues.

Commit:
Merge pull request #42 from k0ekk0ek/relnotes

Add changelog entries and release notes

Commit:
Add regex option to unit test framework

Commit:
Add unit tests for popen3

Commit:
Update changelog and release notes

Commit:
Add unit tests for zone resource record iterator

wcawijngaards committed:
- Fix #44: document that remote-control is a top-level nsd.conf
  attribute.

wcawijngaards committed:
- Fix compile on OSX.

wcawijngaards committed:
- Fix for #44: nicer top-level clause documentation.

Commit:
Add unit tests to verify child processes are reaped

wcawijngaards committed:
- Fix checkconf test with filenames that sort in the same order.

wcawijngaards committed:
- Tag for 4.2.3rc1.

wcawijngaards committed:
- Detect fixed time memcmp for openssl 0.9.8 compatibility.

wcawijngaards committed:
- Detect EC_KEY_new_by_curve_name for openssl 0.9.8.

wcawijngaards committed:
- include limits.h for UINT_MAX.
- If no recvmmsg, dont use msg_flags member, but errno for error,
  where our fallback function left it, msg_flags also does not exist
  on some systems.
- Remove unused variable warning for portability.

wcawijngaards committed:
Branch master is 4.2.4 in development.

Commit:
Add initial Buildkite pipeline.yml

wcawijngaards committed:
- Add make distclean that removes config.h made by configure.

wcawijngaards committed:
- Fix #48: Add make distclean that removes config.h made by configure.
(changelog note, previous commit with code in Makefile).

wcawijngaards committed:
And add maintainer-clean that removes bison and flex output.

wcawijngaards committed:
Fixup clean targets, shorter and dnstap_config.h is deleted too.

Commit:
Merge pull request #49 from k0ekk0ek/buildkite

Add initial Buildkite pipeline.yml

Commit:
Add test to verify minimal-responses works as expected

Commit:
Fix global variables not being set in configparser.y

wcawijngaards committed:
- Fix #52: do not log transient network full errors unless higher
  verbosity is set.

wcawijngaards committed:
- Fix checkconf test for new error output string.

wcawijngaards committed:
- tag for 4.2.4rc1 release.

wcawijngaards committed:
- Fix fname null check of fname in namedb_read_zonefile.

wcawijngaards committed:
- Fix implicit cast of size in udb_radnode_array_grow.

wcawijngaards committed:
- Fix ignore of return value of ssl_printf in remote.c.

wcawijngaards committed:
- Fix unused check of fd in parent_handle_reload_command.

wcawijngaards committed:
- Fix to use getrandom() for randomness, if available.

wcawijngaards committed:
- Attempt to fix signedness of nscount lookup in ixfr query_process.

wcawijngaards committed:
- Fix identical branches for ssl_print of errors in remote.c.

wcawijngaards committed:
- Fix type cast bounds, signedness of opt_rdlen in edns_parse_record.

wcawijngaards committed:
- Fix to separate header and data lines in parse_zone_list_file.

wcawijngaards committed:
- Fix to define max number of EDNS records we are willing to
  spend time on.

wcawijngaards committed:
- Fix size of string len and capacity type cast in udbradtree.

wcawijngaards committed:
- Fix to protect rrcount in tsig_find_rr from overflow.

wcawijngaards committed:
- Annotate radix_find_prefix_node not reachable trail code.

wcawijngaards committed:
- Fix to protect rrcount in packet_find_notify_serial from overflow.

wcawijngaards committed:
- Fix to close socket on error in create_tcp_accept_sock.

wcawijngaards committed:
- Fix to log on failure to chmod for socket for remote control.

wcawijngaards committed:
- Fix to remove unneeded if in open of socket for remote control.

wcawijngaards committed:
- Fix to restore input parameter on call failure in create_dirs.

wcawijngaards committed:
- Please checker by terminating and initialising string read
  by remote control.

wcawijngaards committed:
- Fixup of random_generate negative modulo, from previous commit,
  and return srandom when random is used if no getrandom.

wcawijngaards committed:
- Fix to define upper bounds on rr counts read from untrusted packet
  data.
- Try different annotation for radix_find_prefix_node not reachable.

wcawijngaards committed:
- Separate acl_addr_match_range functions for ip4 and ip6, to
  please checkers.

wcawijngaards committed:
- Avoid unused variable warning in new match_range_v4 function.

wcawijngaards committed:
Add ifdef for test compilation on nonIPv6 systems.

Commit:
Use Ubuntu 18.04 on Travis CI

wcawijngaards committed:
- Fix whitespace in nsd.conf.sample.in, patch from Paul Wouters.

wcawijngaards committed:
- use-systemd is ignored in nsd.conf, when NSD is compiled with
  libsystemd it always signals readiness, if possible.

wcawijngaards committed:
- Note that use-systemd is not necessary and ignored in man page.

wcawijngaards committed:
master branch is 4.2.5 in development.

Commit:
Update mailing lists

wcawijngaards committed:
- Fix #56: Drop sparse TSIG signing support in NSD.
  Sign every axfr packet with TSIG, according to the latest
  draft-ietf-dnsop-rfc2845bis-06, Section 5.3.1.

Commit:
Add cpu affinity and socket partitioning

Commit:
Set process title to identify subprocesses

Commit:
Print server process affinity for sockets on startup

Commit:
Add cpu affinity tests

wcawijngaards committed:
- Fix unreachable code in ssl set options code.

wcawijngaards committed:
Update relnotes.

wcawijngaards committed:
- Fix bad shift in assertion code analyzer complaint.

Commit:
Add invalid server ranges test

buddyns committed:
add FreeBSD support for conf key ip-transparent

wcawijngaards committed:
Merge pull request #59 from buddyns/master

add FreeBSD support for conf key ip-transparent

wcawijngaards committed:
Change note for PR #59.
- Merge pull request #59 from buddyns: add FreeBSD support
  for conf key ip-transparent.

Commit:
Add socket partitioning test

Commit:
Update changelog and release notes

Commit:
Fix checkconf test

wcawijngaards committed:
- Fix leak in server bitset setup.

wcawijngaards committed:
Fix in not released code, no need to put it there.

wcawijngaards committed:
Merge pull request #22 from hdais/min-any

minimise-any: prefer polular and not large RRset

wcawijngaards committed:
Changelog note for PR#22.
- Merge PR#22: minimise-any: prefer polular and not large RRset,
  from Daisuke Higashi.

wcawijngaards committed:
- Fix responses for IXFR so that the authority section is not echoed

wcawijngaards committed:
Merge pull request #60 from michaelforney/misc-fixes

Minor portability fixes

wcawijngaards committed:
Changelog note for #60.
- Merge PR#60: Minor portability fixes from michaelforney, with
  avoid pointer arithmetic on void* and avoid unnecessary VLA.

Commit:
Use wait instead of sleep in xfr_udp to handle timing difference

Commit:
Run all tests not just the socket_partitioning test

Commit:
Use wait instead of sleep in zone_expire_restart to handle timing diffence

k0ekk0ek commented:
Creating a new PR, this is getting too complicated :sweat_smile:
